### PR TITLE
add validator nonce in staking info contract (mat 1109)

### PR DESCRIPTION
**Pull request by AyushKaul**
Only tests for staking and re-staking are added.

### CODE REVIEW

File: contracts/staking/StakingInfo.sol [hunk 1]
```diff
@@ -41,10 +41,12 @@ contract IStakeManager {
 
 contract StakingInfo {
     using SafeMath for uint256;
+    mapping(uint256 => uint256) public validatorNonce;
```
Comment (jdkanani) [inline contracts/staking/StakingInfo.sol:44]: Shouldn't we put this in `StakeManagerStorage`?  Other than everything looks good. 

Also `validatorNonce` in `StakingInfo.sol` is okay 😄

File: contracts/staking/StakingInfo.sol [hunk 2]
```diff
@@ -189,11 +197,13 @@ contract StakingInfo {
         emit Staked(
             signer,
             validatorId,
+            validatorNonce[validatorId].add(1),
             activationEpoch,
             amount,
             total,
             signerPubkey
         );
+        validatorNonce[validatorId] = validatorNonce[validatorId].add(1);
```
Comment (jdkanani) [inline contracts/staking/StakingInfo.sol:206]: We should do it before `emit`, so we don't have to add twice. For all events.

File: contracts/staking/StakingInfo.sol (hunk 1 above)
Comment (AyushKaul) [reply contracts/staking/StakingInfo.sol:44]: `StakeManagerStorage` is used in many other places where the nonce will not make sense. @0xAshish has more context.

File: contracts/staking/StakingInfo.sol (hunk 2 above)
Comment (AyushKaul) [reply contracts/staking/StakingInfo.sol:206]: yes, I'll change it.

Comment (AyushKaul) [reply contracts/staking/StakingInfo.sol]: done

File: test/staking/StakeManager.test.js [hunk 3]
```diff
@@ -146,6 +145,8 @@ contract('StakeManager', async function (accounts) {
 
       logs[1].event.should.equal('Staked')
       logs[1].args.signerPubkey.toLowerCase().should.equal(userPubkey.toLowerCase())
+      // console.log(logs[1].args.nonce)
+      // assertBigNumberEquality(logs[1].args.nonce, 1)
```
Comment (jdkanani) [inline test/staking/StakeManager.test.js:149]: Why removed?

File: contracts/staking/StakingInfo.sol [hunk 4]
```diff
@@ -220,14 +232,16 @@ contract StakingInfo {
...
     }
 
     function logReStaked(uint256 validatorId, uint256 amount, uint256 total)
         public
         onlyStakeManager
-    {
-        emit ReStaked(validatorId, amount, total);
+    {   
+        validatorNonce[validatorId] = validatorNonce[validatorId].add(1);
+        emit ReStaked(validatorId, validatorNonce[validatorId], amount, total);
```
Comment (jdkanani) [inline contracts/staking/StakingInfo.sol:244]: Only the following events need to emit with `nonce` (not any other event):

`Staked`, `StakeUpdate` , `SignerChange` , `UnstakeInit`

File: test/staking/StakeManager.test.js (hunk 3 above)
Comment (AyushKaul) [reply test/staking/StakeManager.test.js:149]: only comments

**Pull request merged**